src/utils/draftbots.js

```diff
@@ -104,29 +104,47 @@ const scaleSimilarity = (value) =>
 const SYNERGY_SCALE = 0.2;
 const MAX_SYNERGY = 10;
 
-let synergyMatrix;
+const xorStrings = (a, b) => {
+  let s = '';
 
-export const setSynergyMatrix = (size) => {
-  synergyMatrix = [];
-  for (let i = 0; i <= size; i++) {
-    synergyMatrix.push(new Array(size + 1).fill(null));
+  // use the longer of the two words to calculate the length of the result
+  for (let i = 0; i < Math.max(a.length, b.length); i++) {
+    // append the result of the char from the code-point that results from
+    // XORing the char codes (or 0 if one string is too short)
+
+    // eslint-disable-next-line no-bitwise
+    s += String.fromCharCode((a.charCodeAt(i) || 0) ^ (b.charCodeAt(i) || 0));
   }
+
+  return s;
 };
 
-export const getSynergy = (index1, index2, synergies) => {
-  if (!synergies) return 0;
-  if (synergyMatrix[index1][index2] === null) {
-    const similarityValue = similarity(synergies[index1], synergies[index2]);
+const synergyCache = {};
+
+export const getSynergy = (card1, card2) => {
+  if (
+    !card1.details.embedding ||
+    !card2.details.embedding ||
+    card1.details.embedding.length === 0 ||
+    card2.details.embedding.length === 0
+  ) {
+    return 0;
+  }
+
+  if (card1.details.name === card2.details.name) {
+    return MAX_SYNERGY;
+  }
+
+  const key = xorStrings(card1.details.name, card2.details.name);
+  if (!synergyCache[key]) {
+    const similarityValue = similarity(card1.details.embedding, card2.details.embedding);
+
     if (Number.isFinite(similarityValue)) {
-      synergyMatrix[index1][index2] = -Math.log(1 - scaleSimilarity(similarityValue)) / SYNERGY_SCALE;
-    }
-    // This happens when similarityValue is 1, usually because a card is a duplicate.
```

Comment:
Proper handling of the degenerate cases is very important here since the positive case can occur if the similarities are close enough and we do want to properly populate a 0 in the other cases.